CRM/Contact/BAO/Contact/Location.php

```diff
@@ -40,16 +40,22 @@ class CRM_Contact_BAO_Contact_Location {
    *
    * @param bool $isPrimary
    * @param int $locationTypeID
+   * @param bool $isBilling
    *
    * @return array
    *   Array of display_name, email, location type and location id if found, or (null,null,null, null)
    */
-  public static function getEmailDetails($id, $isPrimary = TRUE, $locationTypeID = NULL) {
+  public static function getEmailDetails($id, $isPrimary = TRUE, $locationTypeID = NULL, $isBilling = FALSE) {
     $primaryClause = NULL;
     if ($isPrimary) {
       $primaryClause = " AND civicrm_email.is_primary = 1";
     }
 
+    // CRM-20362 retrieving Billing email address.
+    if ($isBilling) {
+      $primaryClause = " AND civicrm_email.is_billing = 1";
```

Comment:
So at the code level - I think the fall back to primary should be happening in the getEmailDetails function not in the calling code and I would be reluctant to merge this without tests. I would probably also wrap that function in a function getBillingEmailDetails rather than just add another input variable.

At the philosophical level - is_billing seems to have been a concept conceived but never really implemented (as summed up in "As that renders the billing email effectively useless.").

The question is then per @seamuslee001 - will there be unforseen circumstances if we start implementing it. I think we should expose is_billing in the UI if we are going to and also consider how is_billing gets set - ie. if someone makes an online donation and they have a home address marked as is_billing is that updated or is a new one created, which steals the is_billing flag.

I think we might need to 'game out' some scenarios as to how is_billing gets set & updated. Maybe email the dev list? (assuming people like @Stoob are on that?)